### ISSUE TITLE
remat unnecessarily forces lazy values

### ISSUE THREAD

**Issue by trevorcai** [google/jax#3143]
```
import jax
import jax.numpy as jnp
import haiku as hk

N_LAYERS = 3

def f(rng, x, y):
  x = hk.dropout(rng, 0.1, x)
  return jax.nn.relu(jnp.matmul(x, y))

k1, k2, k3 = jax.random.split(jax.random.PRNGKey(428), 3)
x = jax.random.uniform(k1, (2, 1024, 1024))
y = jax.random.uniform(k2, (1024, 1024))

def g(rng, x, y):
  for k in jax.random.split(rng, N_LAYERS):
    x = jax.remat(f)(k, x, y)
  return jnp.sum(x)

# Profiling this line.
jax.jit(jax.grad(g, argnums=2))(k3, x, y).block_until_ready()
```

This results in the `lax.iota` from the underlying `jax.random._random_bits` being forced at trace time, rather being evaluated in runtime. With a large enough N_LAYERS or `x.shape`, this can result in OOMing the device _during tracing_.

**Comment by jekbradbury** [comment]
The problem here is in https://github.com/google/jax/blob/master/jax/interpreters/partial_eval.py#L685-L700. We treat `partial_eval_jaxpr` as if it were pure (we just want an analysis result: which outvars depend on which invars?) but it has an unfortunate side effect here: if there are `xla_call` primitives in the jaxpr, their inputs (in this case literals in the jaxpr) are forced (and that memory usage persists). At the very least we'd want to un-force them—after all, this is remat and re-materializing things at runtime is its job!

I'm not really sure what to do now, though. There are changes in flight for both staging/laziness as a whole ("omnistaging") and remat (from Adam), but I'm sure there's also something we could do in the meantime. I'll think about it more and pull Matt in tomorrow.

**Comment by inoryy** [comment]
I've reduced the repro down to the bare minimum:

```python
import jax
import jax.numpy as jnp

def f(y, k): 
  return y * jax.random.uniform(k, y.shape)

def g(x, k):
  z = jax.remat(f)(x, k)
  return jnp.sum(z)

x = jnp.ones((1, 3))
k = jax.random.PRNGKey(42)

jax.jit(jax.grad(g))(x, k).block_until_ready()
```

I've also noticed that if I define `f` inside `g` then the issue goes away, e.g. like so

```python
def g(x, k):
  f = lambda y: y * jax.random.uniform(k, y.shape)  
  z = jax.remat(f)(x)
  return jnp.sum(z)
```

**Comment by trevorcai** [comment]
The main thing I'm not following here is what exactly is causing the force-ing of xla_call inputs during partial_eval_jaxpr. partial_eval_jaxpr seems to delegate to `trace_to_jaxpr(core.jaxpr_as_fun(jaxpr))`, and naively it seems like `trace_to_jaxpr` does the right thing when it's being using by `jit` alone. What does partial_eval_jaxpr change s.t. we now force constants?

**Comment by mattjj** [comment]
Fixed in #3169 !
